org.eclipse.jdt.ls.core/src/org/eclipse/jdt/ls/core/internal/ConnectionStreamFactory.java

```diff
@@ -132,34 +176,41 @@ public OutputStream getOutputStream() throws IOException {
 	 *
 	 * @return
 	 */
-	public StreamProvider getSelectedStream(){
+	public StreamProvider getSelectedStream() {
 		if (provider == null) {
+			final String pipeName = Environment.get("INOUT_PIPE_NAME");
+			if (pipeName != null) {
+				provider = new DuplexPipeStreamProvider(pipeName);
+			}
 			final String stdInName = Environment.get("STDIN_PIPE_NAME");
 			final String stdOutName = Environment.get("STDOUT_PIPE_NAME");
 			if (stdInName != null && stdOutName != null) {
-				provider= new NamedPipeStreamProvider(stdOutName, stdInName);
+				provider = new DualPipeStreamProvider(stdOutName, stdInName);
 			}
-			final String wHost = Environment.get("STDIN_HOST","localhost");
-			final String rHost = Environment.get("STDOUT_HOST","localhost");
+			final String wHost = Environment.get("STDIN_HOST", "localhost");
+			final String rHost = Environment.get("STDOUT_HOST", "localhost");
 			final String wPort = Environment.get("STDIN_PORT");
 			final String rPort = Environment.get("STDOUT_PORT");
 			if (rPort != null && wPort != null) {
 				provider = new SocketStreamProvider(rHost, Integer.parseInt(rPort), wHost, Integer.parseInt(wPort));
 			}
-			if(provider == null ){//Fall back to std io
+			if (provider == null) {//Fall back to std io
 				provider = new StdIOStreamProvider();
 			}
 		}
 		return provider;
 	}
 
-
-	public InputStream getInputStream() throws IOException{
+	public InputStream getInputStream() throws IOException {
 		return getSelectedStream().getInputStream();
 	}
 
-	public OutputStream getOutputStream() throws IOException{
+	public OutputStream getOutputStream() throws IOException {
 		return getSelectedStream().getOutputStream();
 	}
 
+	private boolean isWindows() {
```

Comment:
can we refactor that method some place else? Already exists in DuplexPipeStreamProvider